Batteries/Data/Fin/Lemmas.lean

```diff
@@ -42,63 +42,71 @@ theorem findSome?_succ_of_none {f : Fin (n+1) → Option α} (h : f 0 = none) :
+    findSome? f = some a ↔ ∃ i, f i = some a ∧ ∀ j < i, f j = none := by
   induction n with
-  | zero => rw [findSome?_zero] at h; contradiction
+  | zero =>
+    simp only [findSome?_zero, (Option.some_ne_none _).symm, false_iff]
+    exact fun  ⟨i, _⟩ => i.elim0
   | succ n ih =>
-    rw [findSome?_succ] at h
-    match heq : f 0 with
-    | some x =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_some] at h
-      exists 0
-      rw [heq, h]
-    | none =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_none] at h
-      match ih h with | ⟨i, _⟩ => exists i.succ
+    simp only [findSome?_succ, Option.orElse_eq_orElse, Option.orElse_eq_or, Option.or_eq_some_iff,
+      ih, forall_fin_succ, exists_fin_succ, false_implies, not_lt_zero, implies_true,
+      and_true, succ_pos, forall_const, succ_lt_succ_iff, and_left_comm (b := f 0 = none),
+      exists_and_left]
 
-theorem eq_none_of_findSome?_eq_none {f : Fin n → Option α} (h : findSome? f = none) (i) :
-    f i = none := by
-  induction n with
-  | zero => cases i; contradiction
-  | succ n ih =>
-    rw [findSome?_succ] at h
-    match heq : f 0 with
-    | some x =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_some] at h
-      contradiction
-    | none =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_none] at h
-      cases i using Fin.cases with
-      | zero => exact heq
-      | succ i => exact ih h i
-
-@[simp] theorem findSome?_isSome_iff {f : Fin n → Option α} :
-    (findSome? f).isSome ↔ ∃ i, (f i).isSome := by
-  simp only [Option.isSome_iff_exists]
-  constructor
-  · intro ⟨x, hx⟩
-    match exists_of_findSome?_eq_some hx with
-    | ⟨i, hi⟩ => exists i, x
-  · intro ⟨i, x, hix⟩
-    match h : findSome? f with
-    | some x => exists x
-    | none => rw [eq_none_of_findSome?_eq_none h i] at hix; contradiction
+@[simp]
+theorem findSome?_isSome_iff {f : Fin n → Option α} :
+    (findSome? f).isSome ↔ ∃ i, (f i).isSome ∧ ∀ j < i, (f j).isNone := by
+  simp only [Option.isSome_iff_exists, findSome?_eq_some_iff, exists_comm, exists_and_right,
+    Option.isNone_iff_eq_none]
 
 @[simp] theorem findSome?_eq_none_iff {f : Fin n → Option α} :
     findSome? f = none ↔ ∀ i, f i = none := by
-  constructor
-  · exact eq_none_of_findSome?_eq_none
-  · intro hf
-    match h : findSome? f with
-    | none => rfl
-    | some x =>
-      match exists_of_findSome?_eq_some h with
-      | ⟨i, h⟩ => rw [hf] at h; contradiction
+  induction n with
+  | zero =>
+    simp only [findSome?_zero, true_iff]
+    exact fun i => i.elim0
+  | succ n ih =>
+    simp only [findSome?_succ, Option.orElse_eq_orElse,
+      Option.orElse_eq_or, Option.or_eq_none_iff, ih, forall_fin_succ]
 
 theorem findSome?_isNone_iff {f : Fin n → Option α} :
     (findSome? f).isNone ↔ ∀ i, (f i).isNone := by simp
 
+theorem exists_of_findSome?_eq_some {f : Fin n → Option α} (h : findSome? f = some x) :
+    ∃ i, f i = some x ∧ ∀ j < i, f j = none :=
```

Comment:
Let's split this one in two separate theorems.

Reply:
Can I check, what do you want the two theorems to be?

Reply:
Disregard this split, it doesn't make sense. But let's change the name to `exists_minimal_of_findSome?_eq_some` or something similar.